graphql/src/store/prefetch.rs

```diff
@@ -506,295 +453,309 @@ fn execute_root_selection_set(
-                    }
+// When querying an object type, `type_condition` will always be that object type, even if it passes
+// through fragments for interfaces which that type implements.
+//
+// When querying an interface, `type_condition` will start as the interface itself at the root, and
+// change to an implementing object type if it passes to a fragment with a concrete type condition.
+fn collect_fields_inner<'a>(
+    ctx: &'a ExecutionContext<impl Resolver>,
+    type_condition: ObjectOrInterface<'a>,
+    selection_set: &'a q::SelectionSet,
+    visited_fragments: &mut HashSet<&'a q::Name>,
+    output: &mut IndexMap<&'a String, CollectedResponseKey<'a>>,
+) {
+    fn is_reference_field(
+        schema: &s::Document,
+        object_type: ObjectOrInterface,
+        field: &q::Field,
+    ) -> bool {
+        object_type
+            .field(&field.name)
+            .map(|field_def| sast::get_type_definition_from_field(schema, field_def))
+            .unwrap_or(None)
+            .map(|type_def| match type_def {
+                s::TypeDefinition::Interface(_) | s::TypeDefinition::Object(_) => true,
+                _ => false,
+            })
+            .unwrap_or(false)
+    }
+
+    fn collect_fragment<'a>(
+        ctx: &'a ExecutionContext<impl Resolver>,
+        outer_type_condition: ObjectOrInterface<'a>,
+        frag_ty_condition: Option<&'a q::TypeCondition>,
+        frag_selection_set: &'a q::SelectionSet,
+        visited_fragments: &mut HashSet<&'a q::Name>,
+        output: &mut IndexMap<&'a String, CollectedResponseKey<'a>>,
+    ) {
+        let schema = &ctx.query.schema.document;
+        let fragment_ty = match frag_ty_condition {
+            // Unwrap: Validation ensures this interface exists.
+            Some(q::TypeCondition::On(ty_name)) if outer_type_condition.is_interface() => {
+                schema.object_or_interface(ty_name).unwrap()
+            }
+            _ => outer_type_condition,
+        };
+
+        // The check above makes any type condition on an outer object type redunant.
+        // A type condition on the same interface as the outer one is also redundant.
+        let redundant_condition = fragment_ty.name() == outer_type_condition.name();
+        if redundant_condition || fragment_ty.is_object() {
+            collect_fields_inner(
+                ctx,
+                fragment_ty,
+                &frag_selection_set,
+                visited_fragments,
+                output,
+            );
+        } else {
+            // This is an interface fragment in the root selection for an interface.
+            // We deal with this by expanding the fragment into one fragment for
+            // each type in the intersection between the root interface and the
+            // interface in the fragment type condition.
+            let types_for_interface = ctx.query.schema.types_for_interface();
+            let root_tys = &types_for_interface[outer_type_condition.name()];
```

Comment:
Is it possible that `outer_type_condition` is an object type, but `fragment_ty` is an interface? I think that would lead to a panic here.

Reply:
The comment states that `This is an interface fragment in the root selection for an interface`. It's a bit subtle, but if the outer type is an object, we consider any condition to be redundant and therefore don't enter this if arm.